src/ratbag_emu/protocol/base.py

```diff
@@ -199,32 +197,82 @@ def _send_packets(self, packets, total):
-    def __init__(self, rdesc=None, info=None, name='Generic Device',
-                 shortname='generic'):
-        self.info = info
-        self.name = name
-        self.shortname = shortname
-        self.endpoints = {}
+    @property
+    def shortname(self):
+        return self._owner.shortname
 
-        self.endpoints[0] = SimulatedDevice(self, rdesc, info, name, shortname)
 
-        self.protocol = None
+class BaseDevice(object):
+    device_list = {
+        # Logitech
+        'logitech-g-pro-wireless': LogitechGProWirelessDevice
+    }
+
+    def __init__(self, hw_settings=[], name='Generic Device',
+        info=(0x03, 0x0001, 0x0001), rdescs=[
+            [
+                0x05, 0x01,  # .Usage Page (Generic Desktop)        0
+                0x09, 0x02,  # .Usage (Mouse)                       2
+                0xa1, 0x01,  # .Collection (Application)            4
+                0x09, 0x02,  # ..Usage (Mouse)                      6
+                0xa1, 0x02,  # ..Collection (Logical)               8
+                0x09, 0x01,  # ...Usage (Pointer)                   10
+                0xa1, 0x00,  # ...Collection (Physical)             12
+                0x05, 0x09,  # ....Usage Page (Button)              14
+                0x19, 0x01,  # ....Usage Minimum (1)                16
+                0x29, 0x03,  # ....Usage Maximum (3)                18
+                0x15, 0x00,  # ....Logical Minimum (0)              20
+                0x25, 0x01,  # ....Logical Maximum (1)              22
+                0x75, 0x01,  # ....Report Size (1)                  24
+                0x95, 0x03,  # ....Report Count (3)                 26
+                0x81, 0x02,  # ....Input (Data,Var,Abs)             28
+                0x75, 0x05,  # ....Report Size (5)                  30
+                0x95, 0x01,  # ....Report Count (1)                 32
+                0x81, 0x03,  # ....Input (Cnst,Var,Abs)             34
+                0x05, 0x01,  # ....Usage Page (Generic Desktop)     36
+                0x09, 0x30,  # ....Usage (X)                        38
+                0x09, 0x31,  # ....Usage (Y)                        40
+                0x15, 0x81,  # ....Logical Minimum (-127)           42
+                0x25, 0x7f,  # ....Logical Maximum (127)            44
+                0x75, 0x08,  # ....Report Size (8)                  46
+                0x95, 0x02,  # ....Report Count (2)                 48
+                0x81, 0x06,  # ....Input (Data,Var,Rel)             50
+                0xc0,        # ...End Collection                    52
+                0xc0,        # ..End Collection                     53
+                0xc0,        # .End Collection                      54
+            ]
+        ], shortname='custom-generic-device'):
+        if not self.initialized:
+            self.name = name
+            self.shortname = shortname
+            self.info = info
+            self.rdescs = rdescs
```

Comment:
this seems way too complex for what we want.
The initial code was correct: `__init__(self, hw_settings, name, info, rdescs)` and the in subclasses, you just call `super().__init__(my_local_hw_settings, my_local_name, my_local_info, my_local_rdescs)`

Reply:
Okay, I will do that.